modelling/src/main/java/org/axonframework/modelling/command/StatefulCommandHandlingComponent.java

```diff
@@ -0,0 +1,308 @@
+
+    /**
+     * A definition of a model that can be registered. Used for resolving of the models during execution of the stateful
+     * command handlers that are subscribed to this component. Solely used for internal purposes.
+     */
+    private record ModelDefinition<ID, M>(
+            String name,
+            Class<M> modelClass,
+            Function<CommandMessage<?>, ID> commandIdResolver,
+            ModelLoader<ID, M> loader
+    ) {
+
+    }
+
+    /**
+     * A container for models that either loads models lazily or eagerly. Used during the handling of commands by the
+     * stateful command handlers.
+     * <p>
+     * Any defined model will only be initialized once per command handling process.
+     */
+    private class StatefulCommandModelModelContainer implements ModelContainer {
+
+        private final CommandMessage<?> command;
+        private final ProcessingContext context;
+        private final List<LoadedModelDefinition<?, ?>> loadedModels = new CopyOnWriteArrayList<>();
+
+        /**
+         * Initializes a new {@link ModelContainer} for the given {@code command} and {@code context} that lazily loads
+         * the models.
+         *
+         * @param command The command to load models for
+         * @param context The context in which the command is handled
+         */
+        private StatefulCommandModelModelContainer(CommandMessage<?> command, ProcessingContext context,
+                                                   boolean loadEagerly) {
+            this.command = command;
+            this.context = context;
+
+            if (loadEagerly) {
+                modelDefinitions.forEach(this::eagerLoadModelIfHasIdentifier);
+            }
+        }
+
+        @Override
+        public <T> T modelOf(@Nonnull Class<T> modelType, String name) {
+            T loadedModel = alreadyLoadedModelOf(modelType, name);
+            if (loadedModel != null) {
+                return loadedModel;
+            }
+
+            ModelDefinition<?, T> definition = getModelDefinitionFor(modelType, name);
+            return loadModel(definition);
+        }
+
+        private <ID, T> T loadModel(ModelDefinition<ID, T> definition) {
+            ID id = definition.commandIdResolver().apply(command);
+            if (id == null) {
+                throw new IllegalStateException(
+                        "No identifier found for model of type " + definition.modelClass() + " with name "
+                                + definition.name());
+            }
+            T model = definition.loader().load(id, context).join();
+            loadedModels.add(new LoadedModelDefinition<>(definition.name(), definition.modelClass(), id, model));
+            return model;
+        }
+
+        private <ID, T> void eagerLoadModelIfHasIdentifier(ModelDefinition<ID, T> definition) {
+            ID id = definition.commandIdResolver().apply(command);
+            if (id != null) {
+                loadModel(definition);
```

Comment:
Why we don't throw exception in this case, but wait for accessing the model?

Reply:
Not sure about what this comment is intended for. It's for eager-loading if the identifier is non-null.

I decided to scrap eager loading for now, as the use-case is very slim (see my other comment), and it makes everything more complex.
We can always introduce it later.